spark/v4.0/spark/src/main/java/org/apache/iceberg/spark/source/SparkPositionDeltaWrite.java

```diff
@@ -426,17 +428,35 @@ public DeltaWriter<InternalRow> createWriter(int partitionId, long taskId) {
               .writeProperties(writeProperties)
               .build();
 
+      Function<InternalRow, InternalRow> rowLineageProjector =
+          context.dataSchema() != null
+                  && context.dataSchema().findField(MetadataColumns.ROW_ID.fieldId()) != null
+              ? new ProjectRowLineageFromMetadata()
```

Comment:
I was suggesting to pass `context.dataSchema()` to the ProjectRowLineageFromMetadata and eagerly construct the `ProjectingInternalRow` in the constructor if row_id field exists.

We can also move this if-else condition into the `ProjectRowLineageFromMetadata` class, which can return null if write schema doesn't contain row_id field.

Reply:
ok. looks like `context.dataSchema()` can be null. when could it be null? if it is null, row lineage is not carried over. doesn't it violate the spec?

Reply:
>ok. looks like context.dataSchema() can be null. when could it be null? if it is null, row lineage is not carried over. doesn't it violate the spec?

Yeah I observed that during the very initial analysis rules, context.dataSchema() won't technically be defined at certain points but SparkPositionDeltaWrite will still attempted to be built. Ultimately before execution it'll always be non-null because there will be some output schema for a write. I added the null check to be defensive otherwise we'd fail with an NPE in the middle of analysis needlessly when trying to lookup to determine if lineage fields are defined.


But I do hear your bigger point, which is it's probably cleaner to try and abstract as much as possible behind the project logic